### PR TITLE
add gradient for `compute`

### PR DESCRIPTION
**Pull request by mileslucas**
This will be the WIP PR for adding gradients and jacobians to the limb darkening functions.

I'm still getting used to using ChainRulesCore, so for now I'll try and derive `*_jacobian` or `*_grad` functions and try and add `ChainRulesCore.frule`s for them.

The goal is to be able to get the gradient of the integrated flux with respect to `u_n, b, r`. The first derivative will have to be chain-ruled the way it is in `exoplanet`. The latter two will be written within the main `compute` function.

### CODE REVIEW

**Comment by mileslucas** [comment]
As of right now it seems like the uniform, linear, and quadratic term gradients are consistent with Limbdark.jl based off some plots- no unit tests yet.

So far something I've found interesting is that the jacobian for `dfdu` from `dfdg` is constant for the same `n_max`. Perhaps this could be folded into the computations somehow.

**Comment by mileslucas** [comment]
Okay, all the derivatives seem to be correct based on plots :) next up: some unit tests and figuring how to use ChainRulesCore

**Comment by mileslucas** [comment]
I've got working `frule`s and `rrule`s for the typical gradients, and I've checked they are self-consistent (i.e. calculating grad via forward mode and reverse mode manually were equal).

Also, with the working `rrule`s I've checked that Zygote.jl properly hooks into the chain rules. There's some type stability issues, and I think I've got an idea for how to structure the code going forward.

To fully leverage ChainRules we can play around with `@thunk`ing computations, which will hide some computations behind a closure. So I think we should use this and define everything inside the `rrule`, then rewrite the `compute_grad` function in terms of `rrule` calls and `extern` the output. This should *hopefully* fix some of the type instability inside the rules, then we'll get an idea of what the performance is.

Here's a new plot, too :) 

Screen Shot 2021-02-28 at 10 22 29 PM

File: src/polynomial/poly-grads.jl [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+        ∇flux = ∇flux - ld.g_n[begin + n] * SA[dpdb_M, dpdr_M]
+
+    end
+    dfdg[begin] -= flux * ld.norm * π
+    dfdg[begin + 1] -= flux * ld.norm * π * 2/3
+    dfdb, dfdr = ∇flux * ld.norm
+    return flux * ld.norm, dfdg * ld.norm, dfdb, dfdr
+end
+
+function frule((_, Δld, Δb, Δr), ::typeof(compute), ld::LD, b, r) where {LD<:PolynomialLimbDark}
```
Comment (giordano) [inline src/polynomial/poly-grads.jl:309]: Am I missing something ir LD here and in most other methods below isn't actually used in the body of the function?

Comment (mileslucas) [reply src/polynomial/poly-grads.jl]: https://julialang.slack.com/archives/C6G240ENA/p1617498835147700

**Comment by mileslucas** [comment]
@giordano mind taking a glance for any glaring bugs/errors before I merge?

File: docs/src/api.md [hunk 2]
```diff
@@ -18,6 +18,58 @@ compute
...
+
+# output
+4-element Vector{Float64}:
+  0.0004972185834858653
+ -0.2419262730830416
+ -0.0048107583897073185
+ -0.0024501564976671724
+```
+
+For the most granular support for gradients and jacobians, peer into the depths of `polynomial/poly-grad.jl` and `polynomial/quad-grad.jl`. These functions are not part of the public API and should are not guaranteed any stability according to [semantic versioning](https://semver.org/).
```
Comment (giordano) [inline docs/src/api.md:71]:
```suggestion
For the most granular support for gradients and jacobians, peer into the depths of `polynomial/poly-grad.jl` and `polynomial/quad-grad.jl`. These functions are not part of the public API and are not guaranteed any stability according to [semantic versioning](https://semver.org/).
```

File: src/polynomial/poly-grads.jl [hunk 3]
```diff
@@ -0,0 +1,386 @@
+import ChainRulesCore: frule, rrule
+using LinearAlgebra
+
+function compute_gn_jac(u_n::AbstractVector{T}) where T
+    n = length(u_n) - 1
+    g_n = zero(u_n)
+    dgdu = fill!(similar(u_n, n + 1, n + 1), zero(T))
```
Comment (giordano) [inline src/polynomial/poly-grads.jl:7]: Isn't this something like
```suggestion
    dgdu = zeros(T, n + 1, n + 1)
```
?

File: src/polynomial/poly-grads.jl [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+            bcoeff *= (i - j) / (j + 1)
+        end
+    end
+
+    ## Second: calculate the g_n terms (eqn. 15)
+    for j in n:-1:2
+        if j ≥ n - 1
+            g_n[begin + j] = a_n[begin + j] / (j + 2)
+            for i in 0:n - 1
+                dgdu[begin + i + 1, begin + j] = dadu[begin + i + 1, begin + j] / (j + 2)
```
Comment (giordano) [inline src/polynomial/poly-grads.jl:29]: Isn't it equivalent to
```suggestion
            for i in 1:n
                dgdu[begin + i, begin + j] = dadu[begin + i, begin + j] / (j + 2)
```

File: src/polynomial/poly-grads.jl [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+            end
+        else
+            g_n[begin + j] = a_n[begin + j] / (j + 2) + g_n[begin + j + 2]
+            for i in 0:n - 1
+                dgdu[begin + i + 1, begin + j] = dadu[begin + i + 1, begin + j] / (j + 2) + dgdu[begin + i + 1, begin + j + 2]
```
Comment (giordano) [inline src/polynomial/poly-grads.jl:34]: Same here:
```suggestion
            for i in 1:n
                dgdu[begin + i, begin + j] = dadu[begin + i, begin + j] / (j + 2) + dgdu[begin + i, begin + j + 2]
```

File: src/polynomial/poly-grads.jl [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+            end
+        end
+    end
+    if n ≥ 3
+        g_n[begin + 1] = a_n[begin + 1] + 3 * g_n[begin + 3]
+        for i in 0:n - 1
+            dgdu[begin + i + 1, begin + 1] = dadu[begin + i + 1, begin + 1] + 3 * dgdu[begin + i + 1, begin + 3]
```
Comment (giordano) [inline src/polynomial/poly-grads.jl:41]: And here
```suggestion
        for i in 1:n
            dgdu[begin + i, begin + 1] = dadu[begin + i, begin + 1] + 3 * dgdu[begin + i, begin + 3]
```

File: src/polynomial/poly-grads.jl [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+        end
+    elseif n ≥ 1
+        g_n[begin + 1] = a_n[begin + 1]
+        for i in 0:n - 1
+            dgdu[begin + i + 1, begin + 1] = dadu[begin + i + 1, begin + 1]
```
Comment (giordano) [inline src/polynomial/poly-grads.jl:46]:
```suggestion
        for i in 1:n
            dgdu[begin + i, begin + 1] = dadu[begin + i, begin + 1]
```

File: src/polynomial/poly-grads.jl [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+        end
+    end
+    if n ≥ 2
+        g_n[begin] = a_n[begin] + 2 * g_n[begin + 2]
+        for i in 0:n - 1
+            dgdu[begin + i + 1, begin] = dadu[begin + i + 1, begin] + 2 * dgdu[begin + i + 1, begin + 2]
```
Comment (giordano) [inline src/polynomial/poly-grads.jl:52]:
```suggestion
        for i in 1:n
            dgdu[begin + i, begin] = dadu[begin + i, begin] + 2 * dgdu[begin + i, begin + 2]
```

File: src/polynomial/poly-grads.jl [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+        end
+    else
+        g_n[begin] = a_n[begin]
+        for i in 0:n - 1
+            dgdu[begin + i + 1, begin] = dadu[begin + i + 1, begin]
```
Comment (giordano) [inline src/polynomial/poly-grads.jl:57]:
```suggestion
        for i in 1:n
            dgdu[begin + i, begin] = dadu[begin + i, begin]
```

File: src/polynomial/series.jl [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+    if k2 ≤ 1
+        Nn[begin] = 0.5 * kap0 - k * kc
+        Nn[begin + 1] = 4 / 3 * sqbr * k2 * (2 * Em1mKdm - Eofk)
+    else
+        Nn[begin] = 0.5 * π
+        Nn[begin + 1] = 4 / 3 * sqbr * k * (2 * Eofk - Em1mKdm)
+    end
+
+    return Nn
+end
```
Comment (giordano) [inline src/polynomial/series.jl:210]:
```suggestion
end

```

**Review by giordano — APPROVED** [review]
Looks good overall!

File: src/polynomial/poly-grads.jl (hunk 3 above)
Comment (mileslucas) [reply src/polynomial/poly-grads.jl:7]: for 99% of applications its equivalent, but in the case that we need to retain the array type (e.g. `u_n` is a `CuArray`) the `similar` call is necessary.

Comment (giordano) [reply src/polynomial/poly-grads.jl:7]: I see.  Maybe drop a comment for future readers? :slightly_smiling_face:

**Pull request merged**